onnxruntime/core/framework/tensorprotoutils.cc

```diff
@@ -642,7 +648,13 @@ Status TensorProtoToTensor(const Env& env, const ORTCHAR_T* model_path,
     // load the file
     ORT_RETURN_IF_ERROR(GetFileContent(
         env, external_data_file_path.c_str(), file_offset, raw_data_len,
-        raw_data, deleter_for_file_data.d));
+        raw_data, callback));
+    const DataTypeImpl* const type = DataTypeImpl::TensorTypeFromONNXEnum(tensor_proto.data_type())->GetElementType();
+    TensorShape tensor_shape{tensor_shape_vec};
+    auto p_tensor = std::make_unique<Tensor>(type, tensor_shape, raw_data, OrtMemoryInfo(CPU,
+                                             OrtAllocatorType::OrtDeviceAllocator));
+    tensor = std::move(*p_tensor);
```

Comment:
Previously it looks like the `tensor` parameter had to be pre-allocated, but with this use case that would be unnecessary.

Given that, it may be better not to merge things into one TensorProtoToTensor implementation as the calling semantics are somewhat complicated vs having a shared file local helper for loading external data and 2 separate functions (ExtDataTensorProtoToTensor and TensorProtoToTensor) as that is a lot simpler for a user to understand. 

We should also avoid the allocation of a buffer in the Tensor instance in DeserializeTensorProto if we're doing to directly point to mmap'd external data.